### ISSUE TITLE
Provide argument type when piping in to an anonymous function

### ISSUE THREAD

**Issue by aslilac** [gleam-lang/gleam#2504]
This works fine:

```gleam
  let today =
    birl.now()
    |> birl.get_day()
  let year =
    today.year
    |> int.to_string()
```

But this does not:

```gleam
  let year =
    birl.now()
    |> birl.get_day()
    |> fn (it) { it.year }
    |> int.to_string()
```

You'll get a squiggle under the usage of `it` in the body of the function, because the compiler is failing to deduce the type.

Using that latest of versions of gleam_stdlib and birl off of hex
Using Gleam 0.33

**Comment by lpil** [comment]
Thank you

**Comment by steven-mathew** [comment]
Hi, I wanna tackle this one!

**Comment by sobolevn** [comment]
I started looking at this issue, here's what I found.

First, I saw @lpil's feedback in https://github.com/gleam-lang/gleam/pull/2526#discussion_r1443994440

I added `println!("{:#?}", fun);` between these two lines: https://github.com/gleam-lang/gleam/blob/3128e3bfaf84d9986aacf8d4a784ad6fe9df8c6f/compiler-core/src/type_/expression.rs#L2857-L2858

And wrote these tests:

```gleam
#[test]
fn provide_argument_type_to_anonymous_function1() {
    assert_module_infer!(
        r#"
pub fn main() {
   let z = #(1,2)
   fn(x) { x.0 }(z)
}
"#,
        vec![("main", "fn() -> Int")]
    );
}

#[test]
fn provide_argument_type_to_anonymous_function2() {
    assert_module_infer!(
        r#"
pub fn main() {
   let z = #(1,2)
   fn(x: #(Int, Int)) { x.0 }(z)
}
"#,
        vec![("main", "fn() -> Int")]
    );
}
```

The first test fails, the second one passes.

The first test shows (some parts are removed by me, because they are not needed):

```
Fn {
    typ: Fn {
        args: [
            Var {
                type_: RefCell {
                    value: Link {
                        type_: Tuple {
                            elems: [
                                Named { name: "Int" },
                                Named { name: "Int" },
                            ],
                        },
                    },
                },
            },
        ],
        retrn: Var {
            type_: RefCell {
                value: Unbound {
                    id: 10,
                },
            },
        },
    },
    args: [
        Arg {
            names: Named { name: "x" },
            annotation: None,
            type_: Var {
                type_: RefCell {
                    value: Link {
                        type_: Tuple {
                            elems: [
                                Named { name: "Int" },
                                Named { name: "Int" },
                            ],
                        },
                    },
                },
            },
        },
    ],
}
```

The second test that passes shows:

```
Fn {
    typ: Fn {
        args: [
            Tuple {
                elems: [
                    Named { name: "Int" },
                    Named { name: "Int" },
                ],
            },
        ],
        retrn: Named { name: "Int" },
    },
    args: [
        Arg {
            names: Named { name: "x" },
            annotation: Some(Tuple(...)),
            type_: Tuple {
                elems: [
                    Named { name: "Int" },
                    Named { name: "Int" },
                ],
            },
        },
    ],
}
```

So, basically the difference is rather small. We need to help `self.infer_fn` to infer correct type for this case, using `args` as context. 

Here's what `args` look like in the same place (the same in both cases):

```
[
    CallArg {
        label: None,
        value: Var {
            constructor: ValueConstructor {
                type_: Tuple {
                    elems: [
                        Named { name: "Int" },
                        Named { name: "Int" },
                    ],
                },
            },
            name: "z",
        },
    },
]
```

I will need to investigate why `Tuple` as `type_` works, while the `Var` with the link to the same type does not work.

**Comment by sobolevn** [comment]
Interesting, in case of `fn(x) { x.0 }(z)` it actually changes the type after https://github.com/gleam-lang/gleam/blob/3128e3bfaf84d9986aacf8d4a784ad6fe9df8c6f/compiler-core/src/type_/expression.rs#L2857 call

Before:

```
typ: Fn {
        args: [
            Var {
                type_: RefCell {
                    value: Unbound {
                        id: 9,
                    },
                },
            },
        ],
        retrn: Var {
            type_: RefCell {
                value: Unbound {
                    id: 10,
                },
            },
        },
    },
```

after:

```
typ: Fn {
        args: [
            Var {
                type_: RefCell {
                    value: Link {
                        type_: Tuple {
                            elems: [
                                Named {
                                    publicity: Public,
                                    package: "",
                                    module: "gleam",
                                    name: "Int",
                                    args: [],
                                },
                                Named {
                                    publicity: Public,
                                    package: "",
                                    module: "gleam",
                                    name: "Int",
                                    args: [],
                                },
                            ],
                        },
                    },
                },
            },
        ],
        retrn: Var {
            type_: RefCell {
                value: Unbound {
                    id: 10,
                },
            },
        },
    },
```

**Comment by sobolevn** [comment]
Now when https://github.com/gleam-lang/gleam/pull/3460 is merged, I will work on pipes.
